### ISSUE TITLE
Cloud Controller pre_start.sh script should not restart directly the consul_agent

### ISSUE THREAD

**Issue by keymon** [cloudfoundry/capi-release#34]
## Issue

[cloud_controller_ng pre-start.sh script restarts directly the consul agent by calling its ctl script](https://github.com/cloudfoundry/cloud_controller_ng/blame/master/bosh/jobs/cloud_controller_ng/templates/pre-start.sh.erb#L97), starting it with as `root` user instead of `vcap`. If `consul_agent` gets restarted, it will fail due permissions trying to write on files owned by `root`

## Context

[cloud_controller_ng pre-start.sh script restarts directly the consul agent by calling its ctl script](https://github.com/cloudfoundry/cloud_controller_ng/blame/master/bosh/jobs/cloud_controller_ng/templates/pre-start.sh.erb#L97). This was [added here](https://github.com/cloudfoundry/cloud_controller_ng/commit/b7022b378fc84ba44380e8a3ec20337fbe407db1) as part of #128916909 (I cannot find the story url).

But the [`consul_agent` is meant to run as `vcap:vcap` user and group](https://github.com/cloudfoundry-incubator/consul-release/commit/d940bc786903ec02540bab57f6617a61a3001027) and the monit file specifies that.

If we restart `consul_agent` by calling the ctl directly from cloud_controller pre-start.sh, `consul_agent` will run as root. That will work, but its files (config&data) will be created as `root`, and later the process won't be able to start due permissions after the VM gets restarted or `monit restart consul_agent` is executed.

More info in this conversation https://cloudfoundry.slack.com/archives/bosh-core-dev/p1478184425003326

## Steps to Reproduce and Current result

 * Deploy `cloud_controller` with `consul_agent`.
 * `consul_agent` will be running as root.
 * run `monit restart consul_agent`. It will fail starting because these files are owned by root. `/var/vcap/jobs/consul_agent/config/*.json` and `/var/vcap/data/consul_agent/serf/*`
 * Or restart the VM. `consul_agent` will not start properly.

In general:
 * `monit restart consul_agent`: Starts consul agent as `vcap:vcap`. Would fail if file permissions are wrong.
 * `monit restart all` or `bosh restart api`: will start consul_agent as root, as `cloud_controller_ng` executes the `pre-start.sh`
 

## Expected result

`consul_agent` should not be restarted from cloud controller, and it should always run as whatever its defined in the consul release. Currently is `vcap:vcap`.

## Current result

See "Steps to Reproduce"

## Possible Fix

 * Stop restarting `consul_agent` from cloud_controller. 
 * If you need to wait for `consul_agent` for service discovery of the DB, it is possible to:
    * add retry logic in CC
    * rely on monit restart service feature.
    * wait in the ctl script until `consul_agent` is started
 * Bosh dev team [mention that "bosh need to get on the feature that will start basic services before other jobs. Coming  up"](https://cloudfoundry.slack.com/archives/bosh-core-dev/p1478186344003355)

## Workaround

Run `monit restart all` or `bosh restart api`. Will restart the `consul_agent` as root.

**Comment by keymon** [comment]
Note: This might get complicated to fix in a deployment without human intervention or recreate the api boxes, because the files for `consul_agent` are owned by `root` and they need to be updated:

``` 
chown vcap.vcap /var/vcap/jobs/consul_agent/config/*.json 
chown vcap.vcap /var/vcap/data/consul_agent/serf/*
```

But it feels also wrong that you add logic to fix that...

**Comment by SocalNick** [comment]
This was done so that the cloud controller can talk to the cf-mysql-release switchboard proxy via consul dns. We should fix, but we have to be very careful not to break this capability.

**Comment by keymon** [comment]
Other problem of this is that you are bypassing the [`pre-start.sh` of consul itself](https://github.com/cloudfoundry-incubator/consul-release/blob/master/jobs/consul_agent/templates/pre-start.erb), which (sadly IMO) changes the `resolv.conf` in the system.

In general it is a violation of the encapsulation of the consul agent job.

**Comment by keymon** [comment]
About my previous comment https://github.com/cloudfoundry/capi-release/issues/34#issuecomment-258189472, we are lucky enough. [The `pre-start.sh` of consul agent changes the permissions of its directories to `vcap:vcap`](https://github.com/cloudfoundry-incubator/consul-release/blob/master/jobs/consul_agent/templates/pre-start.erb#L26)

So you don't need to do anything to fix the permissions, consul_agent job will do it and the VM does not need to be recreated.

**Comment by michaelxupiv** [comment]
We just pushed cloudfoundry/cloud_controller_ng@64425c03 to address this issue. Since we need that behavior on our pre start script, we simply run consul's pre-start followed by starting consul as vcap.

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by keymon** [comment]
It is not clear if https://github.com/cloudfoundry/cloud_controller_ng/commit/64425c03 is a temporary solution. 

I insist, IMO one job should not call other job's ctl scripts. This is breaking the encapsulation principles and coupling the capi-release with the consul-release.

Can you explain why you really need this logic? Maybe there are alternatives. For instance:
 * if the issue is that you need the DB up and resolvable, you can add logic on the cloud controller to wait for a given time until it is ready.
 * Not everybody discovers the DB using consul. Actually, I think that is something you do in development environments. If this is the case, you can avoid this logic by only restarting consul if you use it, for instance with a new property. It is still suboptimal, but not so risky.

Also, I would like to ask: what is the plan to replace this?

**Comment by SocalNick** [comment]
@keymon mostly agree, this is not ideal and is hopefully only a temporary solution. @cppforlife has mentioned that he is considering adding a bosh feature where underlying "infrastructure" jobs can be configured to start before other jobs' pre-start scripts.

I'll let @michaelxupiv and @luan comment on your suggested alternatives.

**Comment by luan** [comment]
@keymon thank you the quick response and for challenging this solution. We really appreciate that you're trying to ensure we're doing the right thing here.

I agree with you completely when you say that one job should not call other job's ctl scripts. It is extremely unfortunate that we have to do that and I know we have failed to explain why that's currently the case.

The reason for that is that we currently run DB migrations as part of cloud_controller's pre-start script, we do that because pre-start scripts aren't bound by BOSH's start timeout.

We can't claim that cloud_controller is started before the migrations are fully ran because if we do we run the risk of a second cloud_controller (not the canary) trying to run the same migrations at the same time. We currently have no mechanism to prevent that other than relying on the deployment process that BOSH gives us. For those reasons we can't have cloud_controller wait for the DB to be reachable.

To address you question about wether this is used in development only: I know of at least a few production environments that use cf-mysql and will need consul discovery to reach it. This is already something that Diego and maybe some other components of CF are relying on. Cloud controller just happens to have a more complex startup because it doesn't have the same locking mechanisms of Diego.

And finally, like @SocalNick said, this is not ideal and we hope we can have a better solution coming from BOSH or from the cf-mysql team at some point in the future. I don't personally know exactly how those would work yet but we are well aware that the present solution isn't pretty or perfect.

Does that make things a bit more clear? Happy to answer more questions should you have any.

Thanks,
Luan, CF CAPI Team

**Comment by keymon** [comment]
@luan That is a great explanation. Thank you for sharing it, I do appreciate the effort. It is also good to know that the issue acknowledged and documented here. 

I am fine with waiting for Bosh to provide such solution.

Anyway, as a side note, I would like to suggest some other alternative implementations that we discussed here, at least for reference:

 * Could we call the `monit restart consul_agent`?

 * Did you consider/try using "depends on" in the monit file to declare a dependency on consul_agent? https://mmonit.com/monit/documentation/monit.html

   From what I checked I think that is the consul_agent is not installed, it gets ignored, if not, consul_agent will be restarted first. As a drawback, if consul_agent is restarted, the cloudcontroller will be restarted as well.

 * Or, we could create a specific job for DB migrations of the CC. That job would have a pre_start.sh script which can wait for the domain in consul to be ready.

**Comment by keymon** [comment]
> We can't claim that cloud_controller is started before the migrations are fully ran because if we do we run the risk of a second cloud_controller (not the canary)

This is really interesting! so we should never update more than one api node at the same time.

**Comment by aashah** [comment]
That's what the canary solves really, we rely heavily on how bosh does the deployment (starting with 1 canary in 1 zone before deploying any other nodes of the same job, allowing just one api node to manage db migrations).

To your points:
* We could use monit to restart the consul_agent. Couple problems with that though:
  * monit doesn't call pre-start, pre-start is a bosh thing, so we'd still have to do that manually (for the consul_agent)
  * monit will likely not be on the `PATH`, so we'd have to use the absolute path for it (no big deal, but...)
* `depends on` doesn't solve the problem since pre-start isn't a monit primitive so it'll run before ANY monit commands are even issued.
* The same `depends on` limitation above applies to having a specific migration job with a pre-start.

To summarize, BOSH flow goes as follows:

1. Run all pre-starts
2. `monit start` all jobs

Thanks,
@aashah and @luan, CF CAPI Team
